File: src3/Course.java
import java.util.ArrayList;

public class Course {
    // Attributes
    private ID courseID;
    private String courseName;
    private ArrayList<Course> prerequisiteCourses = new ArrayList<>();
    private ArrayList<CourseSection> courseSections = new ArrayList<>();
    private int courseTerm;
    private double courseCredit;
    private String courseLetterGrade;

    // Constructors
    public Course(ID courseID, String courseName, int courseTerm, double courseCredit, String courseLetterGrade, ArrayList<Course> prerequisiteCourses) {
        this.courseID = courseID;
        this.courseName = courseName;
        this.courseTerm = courseTerm;
        this.courseCredit = courseCredit;
        this.courseLetterGrade = courseLetterGrade;
        this.prerequisiteCourses = prerequisiteCourses != null ? prerequisiteCourses : new ArrayList<>();
    }

    public Course(ID courseID, String courseName, int courseTerm, double courseCredit, ArrayList<Course> prerequisiteCourses) {
        this(courseID, courseName, courseTerm, courseCredit, null, prerequisiteCourses);
    }

    public Course(ID courseID, String courseName, ArrayList<Course> prerequisiteCourses) {
        this(courseID, courseName, 0, 0.0, null, prerequisiteCourses);
    }

    public Course(ID courseID, String courseName) {
        this(courseID, courseName, 0, 0.0, null, null);
    }

    public Course() {
        // Default constructor
    }

    // Methods
    // Add a prerequisite course to the list
    public void addPrerequisiteCourse(Course course) {
        if (course != null) {
            this.prerequisiteCourses.add(course);
        }
    }

    // View all prerequisite courses
    public ArrayList<Course> viewPrerequisiteCourses() {
        return new ArrayList<>(this.prerequisiteCourses); // Return a copy for encapsulation
    }

    // Add a course section
    public void addCourseSection(CourseSection courseSection) {
        if (courseSection != null) {
            this.courseSections.add(courseSection);
        }
    }

    // View all course sections
    public ArrayList<CourseSection> viewCourseSections() {
        return new ArrayList<>(this.courseSections); // Return a copy for encapsulation
    }

    // Getters and Setters
    public ID getCourseID() {
        return courseID;
    }

    public void setCourseID(ID courseID) {
        this.courseID = courseID;
    }

    public String getCourseName() {
        return courseName;
    }

    public void setCourseName(String courseName) {
        this.courseName = courseName;
    }

    public int getCourseTerm() {
        return courseTerm;
    }

    public void setCourseTerm(int courseTerm) {
        this.courseTerm = courseTerm;
    }

    public double getCourseCredit() {
        return courseCredit;
    }

    public void setCourseCredit(double courseCredit) {
        this.courseCredit = courseCredit;
    }

    public String getCourseLetterGrade() {
        return courseLetterGrade;
    }

    public void setCourseLetterGrade(String courseLetterGrade) {
        this.courseLetterGrade = courseLetterGrade;
    }

    public ArrayList<Course> getPrerequisiteCourses() {
        return new ArrayList<>(this.prerequisiteCourses); // Return a copy for encapsulation
    }

    public void setPrerequisiteCourses(ArrayList<Course> prerequisiteCourses) {
        this.prerequisiteCourses = prerequisiteCourses != null ? prerequisiteCourses : new ArrayList<>();
    }

    public ArrayList<CourseSection> getCourseSections() {
        return new ArrayList<>(this.courseSections); // Return a copy for encapsulation
    }

    public void setCourseSections(ArrayList<CourseSection> courseSections) {
        this.courseSections = courseSections != null ? courseSections : new ArrayList<>();
    }
}

File: Iteration3/Python/Course.py
class Course:
    def __init__(self, course_id=None, course_name=None, course_term=None, course_credit=None, course_letter_grade=None, prerequisite_course=None):
        self.course_id = course_id
        self.course_name = course_name
        self.prerequisite_course = prerequisite_course if prerequisite_course else []
        self.course_sections = []
        self.course_term = course_term
        self.course_credit = course_credit
        self.course_letter_grade = course_letter_grade

    def add_prerequisite_course(self, course):
        self.prerequisite_course.append(course)

    def get_course_id(self):
        return self.course_id

    def set_course_id(self, course_id):
        self.course_id = course_id

    def get_course_name(self):
        return self.course_name

    def set_course_name(self, course_name):
        self.course_name = course_name

    def view_prerequisite_courses(self):
        return self.prerequisite_course

    def get_course_term(self):
        return self.course_term

    def set_course_term(self, course_term):
        self.course_term = course_term

    def get_course_credit(self):
        return self.course_credit

    def set_course_credit(self, course_credit):
        self.course_credit = course_credit

    def get_course_letter_grade(self):
        return self.course_letter_grade

    def set_course_letter_grade(self, course_letter_grade):
        self.course_letter_grade = course_letter_grade

    def get_course_sections(self):
        return self.course_sections

    def set_course_sections(self, course_sections):
        self.course_sections = course_sections
